test/lib/ufe/testObject3d.py

```diff
@@ -228,3 +230,154 @@ def testVisibility(self):
+        cylinderPrim = mayaUsd.ufe.ufePathToPrim(ufe.PathString.string(cylinderPath))
+
+        # stage / primSpec
+        stage = mayaUsd.ufe.getStage(str(proxyShapePath))
+        primSpecCapsule = stage.GetEditTarget().GetPrimSpecForScenePath('/Capsule1');
+        primSpecCylinder = stage.GetEditTarget().GetPrimSpecForScenePath('/Cylinder1');
+
+        # select capsule and cylinder prims
+        ufe.GlobalSelection.get().append(capsuleItem)
+        ufe.GlobalSelection.get().append(cylinderItem)
+
+        # hide selected items
+        cmds.hide(cs=True)
+
+        # get the visibility "attribute"
+        capsuleVisibleAttr = capsulePrim.GetAttribute('visibility')
+        cylinderVisibleAttr = cylinderPrim.GetAttribute('visibility')
+
+        # expect the visibility attribute to be 'invisible'
+        self.assertEqual(capsuleVisibleAttr.Get(), 'invisible')
+        self.assertEqual(cylinderVisibleAttr.Get(), 'invisible')
+
+        # visibility "token" must exists now in the USD data model
+        self.assertTrue(bool(primSpecCapsule and UsdGeom.Tokens.visibility in primSpecCapsule.attributes))
+        self.assertTrue(bool(primSpecCylinder and UsdGeom.Tokens.visibility in primSpecCylinder.attributes))
+
+        # undo
+        cmds.undo()
+
+        # expect the visibility attribute to be 'inherited'
+        self.assertEqual(capsuleVisibleAttr.Get(), 'inherited')
+        self.assertEqual(cylinderVisibleAttr.Get(), 'inherited')
+
+        # visibility token must not exists now in the USD data model after undo
+        self.assertFalse(bool(primSpecCapsule and UsdGeom.Tokens.visibility in primSpecCapsule.attributes))
+        self.assertFalse(bool(primSpecCylinder and UsdGeom.Tokens.visibility in primSpecCylinder.attributes))
+
+        # undo
+        cmds.redo()
+
+        # expect the visibility attribute to be 'invisible'
+        self.assertEqual(capsuleVisibleAttr.Get(), 'invisible')
+        self.assertEqual(cylinderVisibleAttr.Get(), 'invisible')
+
+        # visibility "token" must exists now in the USD data model
+        self.assertTrue(bool(primSpecCapsule and UsdGeom.Tokens.visibility in primSpecCapsule.attributes))
+        self.assertTrue(bool(primSpecCylinder and UsdGeom.Tokens.visibility in primSpecCylinder.attributes))
+
+        # hide selected items again
+        cmds.hide(cs=True)
+
+        # right after, call showHidden -all to make everything visible again
+        cmds.showHidden( all=True )
+
+        # expect the visibility attribute to be 'inherited'
+        self.assertEqual(capsuleVisibleAttr.Get(), 'inherited')
+        self.assertEqual(cylinderVisibleAttr.Get(), 'inherited')
+
+        # This time, expect the visibility token to exists in the USD data model
+        self.assertTrue(bool(primSpecCapsule and UsdGeom.Tokens.visibility in primSpecCapsule.attributes))
+        self.assertTrue(bool(primSpecCylinder and UsdGeom.Tokens.visibility in primSpecCylinder.attributes))
+
```

Comment:
Would have been nice to undo / redo showHidden --- should be quick to add and we'd put this one to rest.

Reply:
Good point. Addressed in 585cca0518ea6683c2187cf3661a6bb9268d0761